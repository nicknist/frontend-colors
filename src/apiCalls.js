export const getProjects = () => {
  return fetch('http://backend-colors.herokuapp.com/api/v1/projects')
    .then(response => {
      if (!response.ok) {
        throw Error('Can not find any projects!')
      }
      return response.json();
    })
}

export const getProject = (id) => {
  return fetch(`http://backend-colors.herokuapp.com/api/v1/projects/${id}`)
    .then(response => {
      if(!response.ok) {
        throw Error('Can not find Palette');
      }
      return response.json();
    })
}

export const getPalettes = () => {
  return fetch('http://backend-colors.herokuapp.com/api/v1/palettes')
    .then(response => {
      if (!response.ok) {
        throw Error('Can not find any projects!')
      }
      return response.json();
    })
}

export const getPalette = (id) => {
  return fetch(`http://backend-colors.herokuapp.com/api/v1/palettes/${id}`)
    .then(response => {
      if(!response.ok) {
        throw Error('Can not find Palette');
      }
      return response.json();
    })
}
