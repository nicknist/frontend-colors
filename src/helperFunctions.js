export const randomColor = () => `#${Math.floor(Math.random()*16777215).toString(16)}`;

export const getFiveRandomColors = (colors) => {
  if (colors.length === 0) {
    return [
      { color: randomColor(), locked: false },
      { color: randomColor(), locked: false },
      { color: randomColor(), locked: false },
      { color: randomColor(), locked: false },
      { color: randomColor(), locked: false }]
  }
  let newArray = [];
  colors.forEach(color => {
    if (color.locked) {
      newArray.push({ color: color.color, locked: true });
    } else {
      newArray.push({ color: randomColor(), locked: false });
    }
  })
  return newArray;
}
