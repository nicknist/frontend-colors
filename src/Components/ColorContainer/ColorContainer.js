import React, { useState } from 'react';
import ColorPicker from '../ColorPicker/ColorPicker';
import PaletteContainer from '../PaletteContainer/PaletteContainer';
import { getFiveRandomColors } from '../../helperFunctions';
import { postPalette, deletePalette, postProject } from '../../apiCalls';

const ColorContainer = () => {
  const [colors, setColors] = useState(getFiveRandomColors([]));
  const [palettes, setPalettes] = useState([]);
  const [paletteName, setPaletteName] = useState('');
  const [projectName, setProjectName] = useState('');

  const handleChange = (color, id) => {
    let newColors = [...colors];
    newColors.splice(id, 1, { color: color, locked: true});
    setColors(newColors);
  }

  const lockColor = (colorHex, id, locked) => {
    let newColors = [...colors];
    let lockedColor = { color: colorHex, locked: !locked };
    newColors.splice(id, 1, lockedColor);
    setColors(newColors);
  }

  const addPalette = async () => {
    if (palettes.length === 3) {
      alert('You can only have up to 3 palettes. Please delete one or make them into a Project!');
    } else {
      let newPalette = await postPalette({
        title: paletteName,
        color1: colors[0].color,
        color2: colors[1].color,
        color3: colors[2].color,
        color4: colors[3].color,
        color5: colors[4].color
      });
      let setPalette = {
        id: newPalette.id,
        name: newPalette.title,
        colors: [
          { color: newPalette.color1, locked: false },
          { color: newPalette.color2, locked: false },
          { color: newPalette.color3, locked: false },
          { color: newPalette.color4, locked: false },
          { color: newPalette.color5, locked: false }
        ]
      };
      setPalettes([...palettes, setPalette]);
    }
  }

  const handleDeletePalette = async (name) => {
    let palette = palettes.findIndex(palette => name === palette.name);
    let fullPalette = palettes.find(palette => name === palette.name);
    let array = [...palettes];
    array.splice(palette, 1);
    await deletePalette(fullPalette.id);
    await setPalettes(array);
  }

  const handlePostProject = async () => {
    if (palettes.length !== 3) {
      alert('Please make 3 Palettes for your Project!');
    } else {
      let project = {
        title: projectName,
        palette1_name: palettes[0] ? palettes[0].name : '',
        palette2_name: palettes[1] ? palettes[1].name : '',
        palette3_name: palettes[2] ? palettes[2].name : ''
      };
      await setPalettes([]);
      await postProject(project)
    }
  }

  return (
    <>
      <section className='container'>
        <ColorPicker
          color={colors[0].color}
          locked={colors[0].locked}
          handleChange={handleChange}
          lock={lockColor}
          number={0}
        />
        <ColorPicker
          color={colors[1].color}
          locked={colors[1].locked}
          handleChange={handleChange}
          lock={lockColor}
          number={1}
        />
        <ColorPicker
          color={colors[2].color}
          locked={colors[2].locked}
          handleChange={handleChange}
          lock={lockColor}
          number={2}
        />
        <ColorPicker
          color={colors[3].color}
          locked={colors[3].locked}
          handleChange={handleChange}
          lock={lockColor}
          number={3}
        />
        <ColorPicker
          color={colors[4].color}
          locked={colors[4].locked}
          handleChange={handleChange}
          lock={lockColor}
          number={4}
        />
      </section>
      <button className='randomizer' onClick={() => setColors(getFiveRandomColors(colors))}>RANDOMIZE COLORS</button>
      <input value={paletteName} type='text' className='palette-name' onChange={(e) => setPaletteName(e.target.value)}/>
      <button className='add-palette randomizer' onClick={addPalette}>Add These To a Palette</button>
      { palettes.length === 0 ? "Please add a Palette" :
        <>
          <PaletteContainer palettes={palettes} deletePalette={handleDeletePalette}/>
          <input value={projectName} type='text' className='project-name' onChange={(e) => setProjectName(e.target.value)}/>
          <button className='add-project randomizer' onClick={handlePostProject}>Make These Palettes a Project!</button>
        </>
      }
    </>
  )
}

export default ColorContainer;
