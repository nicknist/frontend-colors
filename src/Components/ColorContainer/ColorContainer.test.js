import React from 'react';
import ColorContainer from './ColorContainer';
import { shallow, mount } from 'enzyme';
import { getFiveRandomColors } from '../../helperFunctions';
import { postPalette, deletePalette, postProject } from '../../apiCalls';
jest.mock('../../helperFunctions');
jest.mock('../../apiCalls')

describe('ColorContainer', () => {
  beforeEach(() => {
    getFiveRandomColors.mockImplementation(() => [
      { color: 'blue', locked: false },
      { color: 'scott', locked: true },
      { color: 'white', locked: false },
      { color: 'fosterBlue', locked: true },
      { color: 'scottPurple', locked: false }
    ]);
    postPalette.mockImplementation(() => Promise.resolve({
      title: 'TEST POST PALETTE',
      color1: '#867CBC',
      color2: '#E2F7EF',
      color3: '#23889A',
      color4: '#A6E508',
      color5: '#D15120'
    }));
    deletePalette.mockImplementation(() => Promise.resolve('Oh Hello'))
    postProject.mockImplementation(() => Promise.resolve({
      title: 'Hot pants',
      project1_name: 'Colins house',
      project2_name: 'Nicks house',
      project3_name: 'Colin and Nicks Crayons'
    }));
  });

  it('should match the snapshot when there are no palettes', () => {
    const wrapper = shallow(<ColorContainer />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot when there are palettes', () => {
    const wrapper = shallow(<ColorContainer />);
    wrapper.find('.add-palette').simulate('click');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('should run getFiveRandomColors when randomizer button is clicked', () => {
    const wrapper = shallow(<ColorContainer />);
    wrapper.find('.randomizer').at(0).simulate('click');
    expect(getFiveRandomColors).toHaveBeenCalled();
  })

  it('should change the state when the palette name input is updated', async () => {
    const wrapper = mount(<ColorContainer />);
    const mockEvent = {
      preventDefault: jest.fn(),
      target: {
        value: 'DID IT'
      }
    }
    wrapper.find('.palette-name').simulate('change', { target: { value: 'waddup' } });

    expect(wrapper.find('.palette-name').prop('value')).toEqual('waddup');
  });
});
