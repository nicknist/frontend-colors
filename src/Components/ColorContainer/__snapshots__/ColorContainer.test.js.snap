// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorContainer should match the snapshot when there are no palettes 1`] = `
<Fragment>
  <section
    className="container"
  >
    <ColorPicker
      color="blue"
      handleChange={[Function]}
      lock={[Function]}
      locked={false}
      number={0}
    />
    <ColorPicker
      color="scott"
      handleChange={[Function]}
      lock={[Function]}
      locked={true}
      number={1}
    />
    <ColorPicker
      color="white"
      handleChange={[Function]}
      lock={[Function]}
      locked={false}
      number={2}
    />
    <ColorPicker
      color="fosterBlue"
      handleChange={[Function]}
      lock={[Function]}
      locked={true}
      number={3}
    />
    <ColorPicker
      color="scottPurple"
      handleChange={[Function]}
      lock={[Function]}
      locked={false}
      number={4}
    />
  </section>
  <button
    className="randomizer"
    onClick={[Function]}
  >
    RANDOMIZE COLORS
  </button>
  <input
    className="palette-name"
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    className="add-palette randomizer"
    onClick={[Function]}
  >
    Add These To a Palette
  </button>
  Please add a Palette
</Fragment>
`;

exports[`ColorContainer should match the snapshot when there are palettes 1`] = `
<Fragment>
  <section
    className="container"
  >
    <ColorPicker
      color="blue"
      handleChange={[Function]}
      lock={[Function]}
      locked={false}
      number={0}
    />
    <ColorPicker
      color="scott"
      handleChange={[Function]}
      lock={[Function]}
      locked={true}
      number={1}
    />
    <ColorPicker
      color="white"
      handleChange={[Function]}
      lock={[Function]}
      locked={false}
      number={2}
    />
    <ColorPicker
      color="fosterBlue"
      handleChange={[Function]}
      lock={[Function]}
      locked={true}
      number={3}
    />
    <ColorPicker
      color="scottPurple"
      handleChange={[Function]}
      lock={[Function]}
      locked={false}
      number={4}
    />
  </section>
  <button
    className="randomizer"
    onClick={[Function]}
  >
    RANDOMIZE COLORS
  </button>
  <input
    className="palette-name"
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    className="add-palette randomizer"
    onClick={[Function]}
  >
    Add These To a Palette
  </button>
  Please add a Palette
</Fragment>
`;
